MudDataGrid: Add methods to manage column visibility and order

ncigula opened the pull request:
I added public methods that allow hiding and showing of columns in the MudDataGrid and changing their order in the UI.
This way anyone wanting to show/hide/reorder the columns based on their specific needs and wants can apply the showing/hiding/reordering on the DataGrid when the page loads by code.
The specific use case would be if a user of the app wants to have their own order of the columns and see only specific columns the app can save this data and on page load read the users settings and apply them to the DataGrid which by default renders with all the columns and in the default order.

ncigulaatvanado committed:
MudDataGrid: Programatically set shown and hidden columns (#11334)

ncigulaatvanado committed:
DataGridTests: Add tests for DataGridShowAndHideProgramaticallyTest

ncigulaatvanado committed:
MudDataGrid: Use async StateHasChanged

ncigulaatvanado committed:
MudDataGrid: Add OrderColumns async method to set the order of the columns

ncigulaatvanado committed:
DataGridShowAndHideProgramaticallyTest: Add description

ncigulaatvanado committed:
DataGridTests: Add tests for column ordering

ncigulaatvanado committed:
DataGridOrderColumnsProgramaticallyTest: Add test viewer for column ordering

ncigulaatvanado committed:
DataGridTests: Fix test assertions

ncigulaatvanado committed:
MudDataGrid.OrderColumnsAsync - Fix xml comment and method performance

ncigulaatvanado committed:
DataGridTests.DataGrid_OrderColumnsAsync - Fix test component model

ncigulaatvanado committed:
MudDataGrid - Hide/ShowColumnsAsync now takes an IEnumerable as a parameter

versile2 commented:
You will need to fix the build-test error before anyone can look at this for you.

ncigulaatvanado committed:
DataGrid - Run dotnet format

versile2 commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -2064,6 +2064,72 @@ internal async Task ShowAllColumnsAsync()
+
+            var hideableRenderedColumns = RenderedColumns
+                .Where(c => c.hideable)
+                .ToDictionary(column => column.Title);
```
Pretty sure this will throw if Titles aren't unique (they aren't required to be). Might need to perform a groupby operation first or something.

versile2 commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -2064,6 +2064,72 @@ internal async Task ShowAllColumnsAsync()
+
+            var hideableRenderedColumns = RenderedColumns
+                .Where(c => c.hideable)
+                .ToDictionary(column => column.Title);
```
Same here, will throw if column titles aren't unique.

versile2 commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -2096,6 +2162,36 @@ private Task ColumnOrderUpdated(MudItemDropInfo<Column<T>> dropItem)
+            if (columnTitles is null || columnTitles.Count == 0)
+                return;
+
+            var columnDict = RenderedColumns.ToDictionary(c => c.Title);
```
Same here, titles aren't required to be unique

versile2 reviewed (changes requested):
I have general concerns on the `OrderColumnsAsync` method, primarily is that it's not respecting the hard and fast columns, such as Hierarchy column/Select column both of which are fairly regular. Especially since neither of those columns has a Title property. 

I think if it were me I would do your AddRange of the columns, and for the remaining cycle through the list and use the AddColumn method which will appropriate place unordered columns.

Also I would like a unit test proving the column title duplicates are okay for all 3 methods please.

Good work overall.

versile2 commented:
> > I have general concerns on the `OrderColumnsAsync` method, primarily is that it's not respecting the hard and fast columns, such as Hierarchy column/Select column both of which are fairly regular. Especially since neither of those columns has a Title property.
> > I think if it were me I would do your AddRange of the columns, and for the remaining cycle through the list and use the AddColumn method which will appropriate place unordered columns.
> > Also I would like a unit test proving the column title duplicates are okay for all 3 methods please.
> > Good work overall.
> 
> Since neither of them have a title i could check which columns have titles and only reorder those and leave the ones without the title either at the beginning or the end. What do you think about this ?

The more I think about it the more I don't want this method in. In a separate PR if you want to create ordering do it this way:
https://github.com/MudBlazor/MudBlazor/pull/11691#pullrequestreview-3073247695

with Parameter State similar to GroupByOrder in Column{T} anyone could re-order a column by just putting an order in the column. It's a slightly bigger project than these modifications though so no-one has picked it up.

ncigulaatvanado committed:
Support programmatic show/hide for duplicate column titles

Enhanced MudDataGrid to handle duplicate column titles when using ShowColumnsAsync and HideColumnsAsync methods. Updated logic to group columns by title and allow operations on all columns with the same title. Added unit tests to validate behavior with duplicate titles.

ncigulaatvanado committed:
DataGrid - Remove the OrderColumnsAsync method as it could introduce too many bugs.

ncigulaatvanado commented:
Its ready for a view i think,
Yes i would like to tackle the ordering issue in a new PR.
Ill ask some question on discord thank you. Ive been meaning to ask here if you wouldnt mind me contacting you there but cancelled that one.
